tests/test_msisdn.py

```diff
@@ -76,3 +99,24 @@ def test_request_code_via_url_query_params(self) -> None:
         sendSMS_mock = self._render_request(request)
         sendSMS_mock.assert_called_once()
         self.assertEqual(channel.code, 200)
+
+    @patch("sydent.http.httpclient.HTTPClient.post_json_maybe_get_json")
+    def test_bad_api_response_raises_exception(self, post_json) -> None:
+        header = FakeHeader({})
+        resp = FakeResponse(code=400, headers=header), {}
+        post_json.return_value = resp
+        self.sydent.run()
+        request, channel = make_request(
+            self.sydent.reactor,
+            "POST",
+            "/_matrix/identity/api/v1/validate/msisdn/requestToken",
+            {
+                "phone_number": "447700900750",
+                "country": "GB",
+                "client_secret": "oursecret",
+                "send_attempt": 0,
+            },
+        )
+        self.assertRaises(
+            Exception, request.render(self.sydent.servlets.msisdnRequestCode)
+        )
```

Comment:
I'm slightly surprised this works. My understanding is that there are two forms of `assertRaises`:
- the function call form here, which will call first its argument and check that doing so raises
- the context manager form, which checks that evaluating an expression raises.

See https://docs.python.org/3/library/unittest.html?highlight=assertraises#unittest.TestCase.assertRaises

I think what's actually happening here is:

1. We evaluate the expression `request.render(self.sydent.servlets.msisdnRequestCode)`
2. Doing so calls `self.sydent.servlets.msisdnRequestCode.render(request)`.
3. We eventually call this line which uses the patch/mock defined above. `resp` ends up being the `FakeResponse` created earlier.

https://github.com/matrix-org/sydent/blob/567eb1a89dbf13c37b857a82d91dac62a484e528/sydent/sms/openmarket.py#L105-L107

4. Because `resp.code == 400`, we raise an Exception here:

https://github.com/matrix-org/sydent/blob/567eb1a89dbf13c37b857a82d91dac62a484e528/sydent/sms/openmarket.py#L121-L129

5. **BUT**, crucially, that gets caught in the second `except` block here:

https://github.com/matrix-org/sydent/blob/d702647d8a6ba351aa90f5756f5f55d74a3f4f9d/sydent/http/servlets/msisdnservlet.py#L101-L121

6. `render_POST` continues execution and returns `None`.
7. `twisted.web.server.Request.render` continues execution and returns `None`.
8. Having evaluated the second argument to `assertRaises`, we then call `self.assertRaises(Exception, None)`.
9. Doing so tries to call `None()`. This raises `TypeError: 'NoneType' object is not callable`.
10. `TypeError` is a subclass of `Exception`, so `assertRaises` sees the behaviour it expects. (C.f. the [stdlib's exception hierarchy](https://docs.python.org/3/library/exceptions.html?highlight=timeouterror#exception-hierarchy).)
11. The test passes.


Sorry, this has become a wall of text. I think there are two key corrections here:

- `assertRaises` in the form used here first evaluates _and then calls_ its second argument. To check that evaluating an expression/or statements raises an exception, use the context manager form.
- `assertRaises` checks that something raises an exception _which is not suppressed_. Any exception that are raised, caught and not reraised (points 4 and 5) are an implementation detail and not part of the externally testable behaviour.

Reply:
(As a sanity check, try making the mock return `200` instead of `400`. I think you'll see that the test still passes and that we are therefore insane.)

I suggest taking a moment to consider what is under test here, and what the test wants to ensure. We agree on the scenario: the openmarket API returns 400.
- What thing are we testing in this circumstance?
- What is the behaviour we want to ensure occurs? (I.e. what can we detect/measure?)

It looks to me like we're testing Sydent's handling of the request as a whole, as if we're a user of the API. The user knows that something has gone wrong only by inspecting the HTTP response. I suggest this test should do so too; I think we can do this by interrogating the `channel`. 

Note: your instinct is good here! It would be very reasonable to test that `sendTextSMS` raises an `Exception` when its call to `post_json_maybe_get_json` returns `400`. It'd make for a great unit test of `sydent.sms.openmarket`. What we have above---making the request and checking the response---is higher level, and tests multiple different layers interacting together. It's more of an [integration or functional test](https://www.atlassian.com/continuous-delivery/software-testing/types-of-software-testing).

Reply:
Thank you for this write-up! I really appreciate it, and it is very helpful/edifying!